### ISSUE TITLE
cunn/DataParallelTable.lua:337: Share not supported for DataParallelTable

### ISSUE THREAD

**Issue by andreamigliorati** [torch/cunn#304]
Hi, I'm trying to speed up my network but I get an error when I try to use nn.DataParallelTable.
This is what I'm trying to do:
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
m1, m2 = **createModel**(8,48), **createModel**(8,48) 
--8 # of GPUs, 48 hidden unit in the last layer
m2:share(m1,'weight', 'bias') ----THE ERROR IS HERE
prl = nn.ParallelTable()
prl:add(m1) 
prl:add(m2)
prl:cuda()
mlp = nn.Sequential()
mlp:add(prl)
mlp:cuda()
crit = nn.CosineEmbeddingCriterion():cuda()
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
Where the bold functions are:

function **createModel**(nGPU,bot)
   local features = nn.Concat(2)
   local fb1 = nn.Sequential() -- branch 1
   
    fb1:add(nn.SpatialConvolution(1,48,3,3,1,1,1,1))       -- 25 --> 27 --> 25
    fb1:add(nn.ReLU(true))               
    fb1:add(nn.SpatialConvolution(48,128,3,3,1,1,1,1))        
    fb1:add(nn.ReLU(true))
    fb1:add(nn.SpatialConvolution(128,192,3,3,1,1,1,1))     
    fb1:add(nn.ReLU(true))
    fb1:add(nn.SpatialConvolution(192,192,3,3,1,1,1,1))      
    fb1:add(nn.ReLU(true))
    fb1:add(nn.SpatialConvolution(192,128,3,3,1,1,1,1))    -- 25 --> 27 --> 25
    fb1:add(nn.ReLU(true))  
    fb1:add(nn.SpatialMaxPooling(2,2,2,2))             -- 25 --> 12x12
    view = 12    
   
   local fb2 = fb1:clone() -- branch 2
   for k,v in ipairs(fb2:findModules('nn.SpatialConvolution')) do
     v:reset() -- reset branch 2's weights
   end
   features:add(fb1) features:add(fb2) features:cuda()   
   --------------the error is at this line-----------
   features = **makeDataParallel**(features, nGPU)

   local classifier = nn.Sequential()    
   classifier:add(nn.View(256*view*view))
   classifier:add(nn.Dropout(0.5))    
   classifier:add(nn.Linear(256*view*view, 4096)) 
   classifier:add(nn.Dropout(0.5))
   classifier:add(nn.Linear(4096, 4096))
   classifier:add(nn.Tanh())
   classifier:add(nn.Linear(4096, bot))
   classifier:add(nn.Tanh())
   classifier:cuda()
   local model = nn.Sequential():add(features):add(classifier)
   return model
end
////////////////////////////////
and
////////////////////////////////////
function **makeDataParallel**(model, nGPU)
   if nGPU > 1 then
      print('converting module to nn.DataParallelTable')
      assert(nGPU <= cutorch.getDeviceCount(), 'number of GPUs less than nGPU specified')
      local model_single = model
      model = nn.DataParallelTable(1)
      for i=1, nGPU do
         cutorch.setDevice(i)
         model:add(model_single:clone():cuda(), i)
      end
   end
   cutorch.setDevice(1)
   return model
end

Do you possibly know where the error is? Sorry but I'm kinda new at this and I cannot find a way to figure it out. Of course I'm figuring out wrong the net structure. Thanks in advance.

**Comment by andreamigliorati** [comment]
I forgot: the error I get is

[C]: in function 'error'
	...a/torch/install/share/lua/5.1/cunn/DataParallelTable.lua:337: in function 'share'
	/home/andrea/torch/install/share/lua/5.1/nn/Container.lua:97: in function 'share'
	main.lua:123: in main chunk
	[C]: at 0x00406670

**Comment by colesbury** [comment]
As the error message suggests, you can't share parameters across GPUs. If you want parameter sharing, structure your network differently, such as by putting the ParallelTable inside the DataParallelTable

**Comment by andreamigliorati** [comment]
Can you provide any hint on how to do that? I'm a total beginner and what you suggest it seems quite obscure to me. Thank you in advance.

**Comment by soumith** [comment]
@andreamigliorati you have your entire network in nn.ParallelTable. And then as an outer container, you put nn.DataParallelTable.

**Comment by Hokie23** [comment]
Hi Soumith, I have the same issue as above. To be able to share the weights across GPUs, should we be using model = nn.ParallelTable instead of nn.Sequential then adding a parallel model?
